docs/architecture/ErrorReporting.md

```diff
@@ -0,0 +1,144 @@
+# Handling and reporting errors
+
+    WARNING: This document describes future plans for error handling.
+    The UI library does not (yet) implement error handling as described here.
+
+This document describes how errors are handled within the ACS UI library and how they are reported to the surrounding application.
+
+## Error sources
+
+There are two categories of errors, and they are handled differently.
+
+1. ACS backend errors: Errors returned by the REST calls to ACS backend services.
+
+   1. These can be transient or permanent (special case: permission error from REST API).
+
+2. UI errors: Programming errors in the UI library / errors returned from other browser APIs. This is a catch-all for all errors not generated from the ACS SDK libraries.
+
+## ACS backend errors
+
+Handling ACS backend errors is challenging when using either the composites or the pure components with the react bindings provided by the ACS UI library:
+
+* ACS backend API calls are asynchronous
+* ACS backend API calls are often initiated outside of the React component lifecycle methods (e.g on interaction handler callbacks or `useEffect` blocks).
+
+But in both cases, all ACS backend API calls are mediated by the (chat and calling) stateful clients. Thus, the stateful client tees errors to a special field in the state:
+
+```ts
+interface CallClientState {
+    // ... other fields ...
+    errors: { [key: SdkClientMethod]: ErrorState[];
+}
+
+type SdkClientMethod = 'CallClient.createCallAgent' | 'CallClient.getDeviceManager'; // ... and many more
```

Comment:
This type mapping should work. I tested that it omits the constructor and private properties.

```suggestion
type ErrorTarget = keyof CallClient | keyof DeviceManager | keyof ...;
```

[TypeScript playground](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAE2AYwDYEMrDm9BnPOAYXVVSNQEtgA7eAbwCg4W4wpKA3dGbAfWSlUAQQDmtGAG5mrdlx79BZAMox0yANbTWbDt15w+STpWTAAsuhrpxUbaxksA9ACoXjli+JYFcdMSEKajoAOg84LwABMABXACMqZHCosEx0AFscQKoJAHkwGEoIGjxkpw9kYrwYKBjkGGgACggCopKAfgAuPxoATwBKe2c3ZLgAFQALbBIyMQk4SkIYvGAEOAa4CasEVGwlVDwwnQjxiBwfAxgpgNnxOgAaNnxCfyIIdPSYmlMeNrGIDS0IhYJB0SikOAQOIAKxQ8HYEBMSDWADMoO84MoACIAaSOOi8AFpxtcZuQcnQsjRITRUL1NuhONhith1IUmTcRHd4JQSmoaGY-PB-IV0sB8awiXAAJoQGJU87AXz+GjAADunKC8151Ssgo2yAurLgqo1ZLmoVGxNl8sE1IQi0gK0hKL8mopMAA5IRkDEoFhKWyuNMhBaYI9tlkyOtSdlgl7SscvCxDUreObuY1+nAxVcIGsbOheeszqnlSb1ZywxLPHBovFEqNomlMg1ATRgasJODUHBiZNsG3aIrQYVSCE4ABVZ3CABefuAbw+Xx+hWK-3bndHPbgaIxkXQ86wTkqy++gjXNEJp-SxRLivLqe346bqSgGUhrSqfZJIdueS-Epd2gPwEAdS8IRaS9CCoQFEEdDB6WsMUawicodDLdNQ0zfpuk4CBKAQIY4Fcdwk1-OAsWAEwzEsaxbAWJYVjWDYthoHZsDFB1-GMUxgEIPA6gmPwfQyYB3zwR50lMdEwAmFlJJ6NY8DAJVASgQ5XwbUwyg8cQYComiLCsGxxKzPCCKIjxGAAX0YRgYF6VT3XjcxgDzBBCAAXjgQFeggV0yS1SkAB9fOAfzXXodBuhoGJ0jicS4Bs6RGEqPk4BAbogo9NyPO8uAACJMMXbCJEKyQgA)